base/MAPL_ISO8601_Time.F90

```diff
@@ -0,0 +1,1292 @@
+         date%year_ = fields%year_
+         date%month_ = fields%month_
+         date%day_ = fields%day_
+         _RETURN(_SUCCESS)
+      else
+         _RETURN(_FAILURE)
+      end if
+   end function construct_ISO8601Date
+
+   function construct_ISO8601Time(isostring, rc) result(time)
+      character(len=*), intent(in) :: isostring
+      integer, intent(inout) :: rc
+      type(ISO8601Time) :: time
+      type(time_fields) :: fields
+      fields = parse_time(trim(adjustl(isostring)))
+      if(fields%is_valid_) then
+         time%hour_ = fields%hour_
+         time%minute_ = fields%minute_
+         time%second_ = fields%second_
+         time%millisecond_ = fields%millisecond_
+         time%timezone_offset_ = fields%timezone_offset_
+         _RETURN(_SUCCESS)
+      else
+         _RETURN(_FAILURE)
+      end if
+   end function construct_ISO8601Time
+
+   function construct_ISO8601DateTime(isostring, rc) result(datetime)
+      character(len=*), intent(in) :: isostring
+      integer, optional, intent(inout) :: rc
+      type(ISO8601DateTime) :: datetime
+      character, parameter :: DELIMITER = TIME_PREFIX
+      type(ISO8601Date) :: date
+      type(ISO8601Time) :: time
+      integer :: status
+      integer :: time_index = 0
+      time_index = index(isostring,TIME_PREFIX)
+      if(time_index > 0) then
+         datetime%date_ = ISO8601Date(isostring(1:time_index-1), _RC)
+         datetime%time_ = ISO8601Time(isostring(time_index:len(isostring)), _RC)
+         _RETURN(_SUCCESS)
+      else
+         _RETURN(_FAILURE)
+      end if
+   end function construct_ISO8601DateTime
+
+   function construct_ISO8601Duration(isostring, imin, imax, rc) result(duration)
+      character(len=*), intent(in) :: isostring
+      integer, intent(in) :: imin
+      integer, intent(in) :: imax
+      integer, optional, intent(out) :: rc
+      type(ISO8601Duration) :: duration
+      integer :: years = -1
+      integer :: months = -1
+      integer :: days = -1
+      integer :: hours = -1
+      integer :: minutes = -1
+      integer :: seconds = -1
+      integer :: istart = -1
+      integer :: istop = -1
+      logical :: date_found = .FALSE.
+      logical :: time_found = .FALSE.
+      logical :: successful = .FALSE.
+      character :: c
+      integer :: pos
+
+      successful = ((imin > 0) .and. (imax <= len(isostring)) .and. &
+         (imin <= imax) .and. (isostring(imin:imin) == 'P')) 
+
+      do while(successful .and. (pos <= imax))
+         successful = .FALSE.
+         pos = pos + 1
+         c = isostring(pos:pos)
+         if(time_found) then
+            select case(c) 
+               case('H')
+                  if(hours >= 0 .or. minutes >= 0 .or. &
+                     seconds >= 0 .or. istart < 1) cycle
+                  hours = read_whole_number_indexed(isostring, istart, istop)
+                  if(hours < 0) cycle
+                  duration%hours_= hours ! DEBUG WDB
+                  istart = 0
```

Comment:
There seems to be lots of problems here.  First, the entire conditional is very long and screams for helper procedures.

But it is also the case that `hours` is a saved variable and would seem to accidentally be positive the second time this procedure is called.    I at least cannot follow the logic.   

Also, istart is given a value at the end of this subblock, and then never referenced again ...